How to show and hide a split view pane with mvvm guide

Nemonek opened the pull request:
Hey, I've been looking around the Avalonia docs for a while and I spotted this guide with written "to do". I kept an eye on it for a while and I decided to write it. So here's my pull requeste. I hope everything is in order.

thevortexcloud commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
```diff
@@ -3,17 +3,304 @@ id: how-to-show-and-hide-a-split-view-pane-with-mvvm
+
+### Implement a converter
+To do this step we will need the char to change as the Pane goes from shown to hidden and viceversa. To do so we need to create a converter function that will return the necessary char to the button when clicked.<br>
+In the `models` folder create a class named SplitViewIconConverter and make it inherit from the IMultiValueConverter interface.<br>
```
Is the folder name actually lower case? I know this is somewhat pedantic but it does actually make a different on things like Linux with case sensitive file names

thevortexcloud commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
````diff
@@ -3,17 +3,304 @@ id: how-to-show-and-hide-a-split-view-pane-with-mvvm
+```
+Second, define how the method will behave based on the value which will be passed to it in the IList object. We will assume that the first value of the list will be the value of `IsSplitViewPaneOpen`, and will treat it accordingly. When the value is true the method will return '>', otherwise it will return '<'.
+``` C#
+public object? Convert(IList<object?> values, Type targetType, object? parameter, CultureInfo culture)
````
You do not need a multi converter for this. In fact, the code would be simpler with just a simple `IValueConverter`.

https://docs.avaloniaui.net/docs/guides/data-binding/how-to-create-a-custom-data-binding-converter

Nemonek replied:
I experimented a little, and yes, that's right. I used a multi converter due to the fact I didn't know about the usage of Path withe converters!

thevortexcloud commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
```diff
@@ -3,17 +3,304 @@ id: how-to-show-and-hide-a-split-view-pane-with-mvvm
+<img className="screenshot-full" src={SecondExample} alt="Second example" />
 
+## Adding content to the button
+Finally, you now just need to add some content to the button that indicates whether you need to show or hide the SplitView's pane. In this guide we will be using the following characters:
```
Does the button really need to have it's content changed? It seems somewhat superfluous to the main point of the guide, which is showing how to control the open/close state of the split view.

Nemonek replied:
I think it's something people that looks around for this guide might want to do.

thevortexcloud replied:
I think it should be its own separate section then as an "advanced" thing. Otherwise it's just conflating two different concepts together.

Nemonek replied:
By separate section do you mean in the same document or another one?

thevortexcloud replied:
Either is fine with me. One of the Avalonia devs may say something else.

thevortexcloud commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
```diff
@@ -3,17 +3,304 @@ id: how-to-show-and-hide-a-split-view-pane-with-mvvm
+### Implement a converter
+To do this step we will need the char to change as the Pane goes from shown to hidden and viceversa. To do so we need to create a converter function that will return the necessary char to the button when clicked.<br>
+In the `models` folder create a class named SplitViewIconConverter and make it inherit from the IMultiValueConverter interface.<br>
+:::info
+I will use a MultiValueConverter because we need it to be called whenever the IsSplitViewPaneOpen value is changed. (For more information on how IMultiValueConverter behaves see [here](https://docs.avaloniaui.net/docs/guides/data-binding/how-to-bind-multiple-properties))  
```
Avoid first person pronouns

thevortexcloud commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
````diff
@@ -3,17 +3,304 @@ id: how-to-show-and-hide-a-split-view-pane-with-mvvm
+### Bind the converter to the button content.
+Lastly, all that's left to do is bind the converter to the button's content. To do so import the namespace of the converter in the MainWindow.axaml file by adding the following line
+``` xml
+xmlns:convs="clr-namespace:how_to_show_and_hide_a_split_view_pane_with_mvvm.Models"
````
This is the first mention of the namespace. Anybody following the guide up until this point will likely have a different namespace

Nemonek replied:
Do you think it's bette if I put something like `"clr-namespace:YourNamespace.Models"`?

thevortexcloud replied:
You can do that as long as you mention to replace it. Alternatively, you can also show your namespace in other places in the guide so if people start blindly copying and pasting it will also show up.

thevortexcloud commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
````diff
@@ -3,17 +3,304 @@ id: how-to-show-and-hide-a-split-view-pane-with-mvvm
+### Bind the converter to the button content.
+Lastly, all that's left to do is bind the converter to the button's content. To do so import the namespace of the converter in the MainWindow.axaml file by adding the following line
+``` xml
+xmlns:convs="clr-namespace:how_to_show_and_hide_a_split_view_pane_with_mvvm.Models"
````
Please use standard .NET namespace conventions. Meaning dot separated instead of spaces/underscores

Nemonek replied:
Sorry, I didn't know about this. The namespace has been named automatically by Visual Studio. I will correct it tomorrow, as it's getting kinda late for me 😅

thevortexcloud replied:
Visual Studio by default will replace spaces with with underscores for namespaces/project names. But it's an easy thing to fix.

thevortexcloud commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
````diff
@@ -3,17 +3,304 @@ id: how-to-show-and-hide-a-split-view-pane-with-mvvm
+</Window>
+```
+:::warning
+If you want to approach the problem by using just a normal converter and passing the IsSplitViewPaneOpen property to it as parameter, don't try to do that since ConverterParameter does not currently support binding. 
````
The converter itself definitely supports bindings. I am not sure what the issue is here?

Nemonek replied:
Converter yes, but not the parameter. From what I found online, specifically [here](https://github.com/AvaloniaUI/Avalonia/issues/6840), while Converter can be binded, ConverterParameter can't

thevortexcloud commented:
I have added some things I think should be addressed. But to be clear I am not an Avalonia dev so it's also not really up to me either.

Nemonek commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
```diff
@@ -3,17 +3,304 @@ id: how-to-show-and-hide-a-split-view-pane-with-mvvm
+- MainWindow.axaml
+- MainWindowViewModel.cs
 
+Those are the files that you will be editing, except for the one that you will be creating during this guide.
```
Yes, it sounded weird to me too; I modified it so that it's more readable and easy to understand. Or at leat I hope so!

Nemonek commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
```diff
@@ -3,17 +3,304 @@ id: how-to-show-and-hide-a-split-view-pane-with-mvvm
+Those are the files that you will be editing, except for the one that you will be creating during this guide.
+
+:::info
+Although here I will be using the MainWindow.axaml and MainWindowViewModel.cs files, this guide will work for every other View and its respective ViewModel.
```
I meant Views and ViewModels that someone may have/create in the app. I edited it so that it specify.
`Although in this guide we will be using the MainWindow.axaml and MainWindowViewModel.cs files, this guide will work for every additional View and ViewModel you may create.`

Nemonek commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
```diff
@@ -3,17 +3,304 @@ id: how-to-show-and-hide-a-split-view-pane-with-mvvm
+:::
+
+## Adding the SplitView panel to the MainWindow
+The starting point will be something like this:
```
I used what you suggested. It's better phrased than anything I can think off!

Nemonek commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
```diff
@@ -3,17 +3,304 @@ id: how-to-show-and-hide-a-split-view-pane-with-mvvm
+<img className="screenshot-full" src={FirstExample} alt="First example" />
+
+## Adding the SplitView's pane
+Now that we have the SplitView we need to create its pane, which is the part that will be hidden when the split view is closed.
```
Do you mean something like this?
`Now that we have the SplitView we need to access its Pane property and set it's content. The Pane contains the elements that will be hidden when the split view is closed.
To achieve this, use the following code:
/* code */`

thevortexcloud replied:
That's the gist of it. Although I would probably phrase it like something like this:

> A SplitView contains a collapsible area called a Pane, that can have its visibility controlled depending on if the Pane is open or not. To set the content of the Pane, the Pane property must be accessed. This can be done by using the following XAML:

Nemonek replied:
That's way better than mine! Thank you!

Nemonek commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
````diff
@@ -3,17 +3,304 @@ id: how-to-show-and-hide-a-split-view-pane-with-mvvm
+    </SplitView.Pane>
+</SplitView>
+```
+We will now add some additional elements to the pane. First a StackPanel, then a button and a textbox with the text "Settings". The following configurations will ensure that the button and the text are aligned.
````
When specifyng a size closer to the control the size specified in the theme or in a style gets overridden: 

> The styling system implements cascading styles by searching the [logical tree](https://docs.avaloniaui.net/docs/concepts/control-trees) upwards from a control, during the selection step. This means styles defined at the highest level of the application (the App.axaml file) can be used anywhere in an application, but may still be overridden closer to a control (for example in a window, or user control).

[Link to the docs page](https://docs.avaloniaui.net/docs/basics/user-interface/styling/styles#how-it-works)

thevortexcloud replied:
I get that, but the theme will still change everything around it. Which may still lead to things not lining up.

Nemonek commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
```diff
@@ -3,17 +3,304 @@ id: how-to-show-and-hide-a-split-view-pane-with-mvvm
+<SplitView PanePlacement="Right" DisplayMode="CompactInline">
+    <SplitView.Pane>
+        <StackPanel VerticalAlignment="Top" Margin="5" Orientation="Horizontal">
+            <Button>
```
Yes, I know. I kept it open to avoid causing confusion to someone who has never seen it written like that. The Button gets used later when adding content to it.

Nemonek commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
````diff
@@ -3,17 +3,304 @@ id: how-to-show-and-hide-a-split-view-pane-with-mvvm
+```
+
+### Adding some content to the SplitView
+To make things more clear, I will add a TextBlock in the main part of the SplitView, which will be the content you may add and that will always be visible.
````
Edited like this: `To make things more clear, we will add a TextBlock in the main part of the SplitView, which you can replace with other content in the future. This part of the content will always be visible.`

Nemonek commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
````diff
@@ -3,17 +3,304 @@ id: how-to-show-and-hide-a-split-view-pane-with-mvvm
+```
+
+### Adding some content to the SplitView
+To make things more clear, I will add a TextBlock in the main part of the SplitView, which will be the content you may add and that will always be visible.
````
Good idea. I rephrased as follow: `To make things more clear, we will add a TextBlock in the main part of the SplitView, outside the Pane. In the future you may replace the TextBlock with the content you need. This part of the SplitView will always be visible.`. Do you think it's better?

thevortexcloud replied:
I would remove "To make things more clear". It's not needed and sounds very casual.

Nemonek commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
````diff
@@ -3,17 +3,304 @@ id: how-to-show-and-hide-a-split-view-pane-with-mvvm
+```
+
+## Binding the SplitView's `IsPaneOpen` property to a ViewModel's property.
+Fist of all, create a boolean property with a private backing field in the ViewModel which will contains the boolean value. Now, by using Reactive UI, make it notify every change to the field to the UI. Remember to initialize the backing property with the value you need in the VewModel's constructor.
````
I'll work on a different implementation and will include it in the guide I think tomorrow. I used Reactview UI because I often saw it being used in the docs, but I will add a paragraph to implement this using the INotifyPropertyChanged interface!

Nemonek commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
````diff
@@ -3,17 +3,304 @@ id: how-to-show-and-hide-a-split-view-pane-with-mvvm
+```
+
+## Binding the SplitView's `IsPaneOpen` property to a ViewModel's property.
+Fist of all, create a boolean property with a private backing field in the ViewModel which will contains the boolean value. Now, by using Reactive UI, make it notify every change to the field to the UI. Remember to initialize the backing property with the value you need in the VewModel's constructor.
````
Can I ask why? English is not my first language and I tought it would have been okay to use it 😅

thevortexcloud replied:
It just sounds very casual. Typically with technical documents/instructions you avoid a casual tone as you are giving directions, not trying to have a conversation with the reader. Instead you use a neural tone that makes statements of fact. It also often results in more explicit instructions that are less ambiguous. EG an example of the difference can be seen below:

"To get to the park, you should probably take a left then a right at the next street"

Compared to:

"The park can be reached by taking the first left street, and then the subsequent first right street."

Nemonek replied:
Got it. Thank you!

Nemonek commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
```diff
@@ -3,17 +3,304 @@ id: how-to-show-and-hide-a-split-view-pane-with-mvvm
+Fist of all, create a boolean property with a private backing field in the ViewModel which will contains the boolean value. Now, by using Reactive UI, make it notify every change to the field to the UI. Remember to initialize the backing property with the value you need in the VewModel's constructor.
 :::info
-This technique uses a complex **binding path** to locate the parent view model for the 
+By default in an MVVM project each ViewModel inherits from a ViewModelBase which inherits from the ReactiveObject class. By doing so in every ViewModel there will be the necessary methods to notify the UI. For more informations see [here](https://docs.avaloniaui.net/docs/concepts/reactiveui/reactive-view-model).
```
Right! I'll fix this when I'll add the part using INotifyPropertyChanged.

Nemonek commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
```diff
@@ -3,17 +3,304 @@ id: how-to-show-and-hide-a-split-view-pane-with-mvvm
-<img src={SplitViewExpanderScreenshot} alt=""/>
+    public MainWindowViewModel()
+    {
+        this._isSplitViewPaneOpen = false;
```
I disagree. I think that for someone who doesn't have much experience with C# it's better to directly see the dafult value being assigned. I added that line also because someone may need the Pane to be shown when starting the app/visualizing the view, and for someone new to Avalonia may be better to know what to modify from the start

thevortexcloud replied:
I would suggest then adding a comment in the code explaining that

Nemonek replied:
Done!

Nemonek commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
````diff
@@ -3,17 +3,304 @@ id: how-to-show-and-hide-a-split-view-pane-with-mvvm
+    }
+}
+```
+Now, to make the button invoke the command you just need to bind it. (For more informations see [here](https://docs.avaloniaui.net/docs/0.10.x/data-binding/binding-to-commands))
````
Didn't notice that! When I'll add the part for INotifyPropertyChanged I will also find a way to add a part where I briefly explain how to do that without ReactiveUI

Nemonek commented:
> I have added some things I think should be addressed. But to be clear I am not an Avalonia dev so it's also not really up to me either.

Hey, thank you so much for the time! This is my fisrt time trying to contribute to a repository, and every advice or constructive criticism is always welcome!
I already fixed/changed the most of what you addressend, except for the major things that I will fix tomorrow.

thevortexcloud commented:
> Just one question: to make a framework agnostic approach do you think it's ok to have a section with Using ReactiveUI and another with "Using INotifypropertyChanged"?

That's fine with me.

thevortexcloud commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
```diff
@@ -9,24 +9,23 @@ import PaneHidden from '/img/guides/development-guides/How-to-show-and-hide-a-sp
 - MainWindowViewModel.cs
 
-Those are the files that you will be editing, except for the one that you will be creating during this guide.
+Those are the files you will be editing the most. (Please note that additional files may need to be created as you proceed through this guide.)
```
Typically you use `()` in the middle of sentence. Not after one.

thevortexcloud commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
```diff
@@ -9,24 +9,23 @@ import PaneHidden from '/img/guides/development-guides/How-to-show-and-hide-a-sp
 
 ## Starting
 First of all, you need to create an MVVM application project, which you can do by reading the following guide [Create and run a project](https://docs.avaloniaui.net/docs/get-started/test-drive/create-a-project).<br>
 Once you created the project you should be presented with the following files:
```
I would probably rephrase this to: "The default MVVM project template will generate a project populated with the following default files:"

thevortexcloud commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
```diff
@@ -52,7 +51,7 @@ The starting point will be something like this:
 For the sake of simplicity we will be deleting the default TextBlock and its value in the MainWindowViewModel.
 <br>
-Now, create the panel and define the basics of how you want it to behave (for more informations on how this control works see [here](https://docs.avaloniaui.net/docs/reference/controls/splitview)) 
+Now, create the SplitView and define the basics of how you want it to behave (for more informations on how this control works see [here](https://docs.avaloniaui.net/docs/reference/controls/splitview)) 
```
It should be "information" not "informations"

thevortexcloud commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
```diff
@@ -83,7 +82,7 @@ We will now add some additional elements to the pane. First a StackPanel, then a
 
 ### Adding some content to the SplitView
-To make things more clear, I will add a TextBlock in the main part of the SplitView, which will be the content you may add and that will always be visible.
+To make things more clear, we will add a TextBlock in the main part of the SplitView, outside the Pane. In the future you may replace the TextBlock with the content you need. This part of the SplitView will always be visible.
```
Remove "the" from "In the future" and add a comma after "future"

Nemonek replied:
Sorry, but I don't understand the difference. What I used sounds right to me

thevortexcloud commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
```diff
@@ -184,9 +183,9 @@ Finally, you now just need to add some content to the button that indicates whet
 
 ### Implement a converter
 To do this step we will need the char to change as the Pane goes from shown to hidden and viceversa. To do so we need to create a converter function that will return the necessary char to the button when clicked.<br>
```
Vice versa is two words.

timunie commented:
Hello @Nemonek 

first of all thank you for the contribution. However I have one idea to make this sample even easier to explore. We have https://github.com/AvaloniaUI/Avalonia.Samples where we want to create tutorials and detialed samples. These can be linked from the docs. Benefit is that anyone can just grab the source, download it and play around. From what I see your sample would fit perfectly here. What do you think?

Nemonek commented:
Alright, so, I fixed some things that were to be done, and I also added the parts with the framework agnosti approach as suggested.
@timunie I think it would be a good idea and I'd love to do so! Just some questions though, as I said I know have the code with both Reactive UI and without. Should I upload the one with Reactive UI, the one without or both?
Also, should I rename the project from How to show and hide ecc... to like "Show and hide a split view pane with MVVM"?

timunie commented:
I think both would be fine. Make it a draft PR for now as I may want to feedback about the sample beforehand. I'm thinking about a Hamburger Menu like solution for the demo, could be under custom controls. 

What I don't understand is why we use a button instead of a togglebutton. A toggleButton can just be bound to IsChecked and we are there without much hassle. This is why I want to see your sample life to understand what may improve.

Nemonek commented:
Ok, after playing a bit with Github desktop I managed to fork the repository. Anyway, I uploaded both the projects in the same directory as the main solution, as I don't know if you want to put those in the MVVM folder or another. Here's the [link](https://github.com/AvaloniaUI/Avalonia.Samples/pull/99/files).
I also renamed both projects, as I firstly copy pasted the project with reactive UI on my pc and just removed it from the copy.

For what regards the toggle button, well, I hadn't tought of that; sure it would be better and especially simpler to use that, as there wouldn't even be the need for a command. Just a simple binding. Let me know what should be modified and I'll take care of it!

Nemonek commented:
@thevortexcloud By separate section do you mean in the same document or another one?

Nemonek commented on docs/guides/development-guides/how-to-show-and-hide-a-split-view-pane-with-mvvm.md:
````diff
@@ -3,17 +3,304 @@ id: how-to-show-and-hide-a-split-view-pane-with-mvvm
+public ICommand ChangeSplitViewPaneStatusCommand { get; }
+```
+And initialize it in the constructor.
+``` C#
````
I think it would bring some confusion to someone new, seeing everything already initialized and done

MikeCodesDotNET commented:
I've just tried running this locally, and perhaps it's my config, but it was full of issues (it kept crashing Docusaurus). If someone can sanity check that it at least works, I'll spend some more time looking at it.

Nemonek commented:
If you need to know if the example works, yes, it did when I created the branch and there's the code in a branch on Avalonia.Samples ( If needed I can upgrade the code to the latest Avalonia's version and check if it still works as it should). If you need to know if it works with docusaurus I'll try to install and run everything as soon as I get home in about 40/60 min and if it doesn't I'll try my best to understand why ( I never used docusaurus so I can't assure anything ).

timunie commented:
Hey @Nemonek , 

many thanks for all your hard work that went into this PR. It is really appreciated. However, we think that this is a bit too specialized to go into the docs the way it is. 

Below is a snippet that shows all what is needed to make it work: 

```xml
<SplitView PaneBackground="Aqua" DisplayMode="CompactOverlay" IsPaneOpen="{Binding IsPaneOpen}">
    <SplitView.Pane>
        <DockPanel>
            <ToggleButton DockPanel.Dock="Top" IsChecked="{Binding $parent[SplitView].IsPaneOpen}" 
                          Content="+" />
            <TextBlock Text="Other Content"></TextBlock>
        </DockPanel>
    </SplitView.Pane>
</SplitView>
```

> [!NOTE]: 
> We already have other places in the docs explaining the MVVM pattern and co, so no need to repeat it here. 

I will close this PR for the above reasons. Hope you can understand this (hard) decision. 


All the best

Tim
